contracts/lockdrop/src/contract.rs

```diff
@@ -392,8 +455,427 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+            AssetInfo::NativeToken { denom } if denom != UNTRN_DENOM => Some(denom.clone()),
+            _ => None,
+        })
+        .ok_or_else(|| StdError::generic_err("No second leg of pair found"))?;
+    attrs.push(attr("token_denom", token_denom.clone()));
+    let token_balance = deps
+        .querier
+        .query_balance(&env.contract.address, token_denom.as_str())?
+        .amount;
+    attrs.push(attr("token_balance", token_balance.to_string()));
+
+    let base = Asset {
+        amount: ntrn_to_new_pool,
+        info: AssetInfo::NativeToken {
+            denom: UNTRN_DENOM.to_string(),
+        },
+    };
+    let other = Asset {
+        amount: token_balance,
+        info: AssetInfo::NativeToken {
+            denom: token_denom.to_string(),
+        },
+    };
+    let mut funds = vec![
+        Coin {
+            denom: UNTRN_DENOM.to_string(),
+            amount: ntrn_to_new_pool,
+        },
+        Coin {
+            denom: token_denom,
+            amount: token_balance,
+        },
+    ];
+
+    funds.sort_by(|a, b| a.denom.cmp(&b.denom));
+    msgs.push(CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: new_pool_addr,
+        funds,
+        msg: to_binary(&astroport::pair::ExecuteMsg::ProvideLiquidity {
+            assets: vec![base, other],
+            slippage_tolerance,
+            auto_stake: None,
+            receiver: None,
+        })?,
+    }));
+
+    attrs.push(attr("provide_liquidity", "true"));
+
+    msgs.push(CallbackMsg::MigratePairStep3 { pool_type }.to_cosmos_msg(&env)?);
+
+    Ok(Response::new().add_messages(msgs).add_attributes(attrs))
+}
+
+fn migrate_pair_step_3(
+    deps: DepsMut,
+    _info: MessageInfo,
+    env: Env,
+    pool_type: PoolType,
+) -> StdResult<Response> {
+    let config: Config = CONFIG.load(deps.storage)?;
+    let mut attrs = vec![
+        attr("action", "migrate_pair_step_3"),
+        attr("pool_type", pool_type),
+    ];
+    let mut new_pool_info: PoolInfoV2 = ASSET_POOLS_V2.load(deps.storage, pool_type)?;
+    // get current balance
+    let balance_response: cw20::BalanceResponse = deps.querier.query_wasm_smart(
+        &new_pool_info.lp_token,
+        &Cw20QueryMsg::Balance {
+            address: env.contract.address.to_string(),
+        },
+    )?;
+    let current_balance = balance_response.balance;
+    attrs.push(attr("current_balance", current_balance.to_string()));
+
+    //update pool info
+    new_pool_info.amount_in_lockups = current_balance;
+    ASSET_POOLS_V2.save(deps.storage, pool_type, &new_pool_info)?;
+
+    // send stake message to generator
+    let stake_msgs = stake_messages(
+        config,
+        env.block.height + 1u64,
+        new_pool_info.lp_token,
+        current_balance,
+    )?;
+
+    MIGRATION_STATUS.save(deps.storage, &MigrationState::MigrateUsers(0u64))?;
+
+    Ok(Response::new()
+        .add_messages(stake_msgs)
+        .add_attributes(attrs))
+}
+
+fn migrate_users(
```

Comment:
Don't you want to simplify code of this function and move out contents of different blocks (`if`, `match` etc) into their own functions, this will simplify code and increase its readability.